kyo-prelude/shared/src/main/scala/kyo2/Result.scala

```diff
@@ -4,134 +4,159 @@ import Result.*
 
-        // TODO avoid Option allocation
-        def unapply[A](self: Result[A]): Option[Throwable] =
-            self.fold(Some(_))(_ => None)
+    object Failure:
+        def unapply[E, A](result: Result[E, A]): Option[E] =
+            (result: @unchecked) match
+                case result: Failure[E] => Some(result.error)
+                case _                  => None
     end Failure
 
-    extension [A](self: Result[A])
-        inline def isSuccess: Boolean =
+    case class Panic(exception: Throwable) extends Cause[Nothing]
+
+    extension [E, T](self: Result[E, T])
+
+        def isSuccess: Boolean =
             self match
-                case _: Failure[?] => false
-                case _             => true
+                case _: Cause[?] => false
+                case _           => true
 
-        inline def isFailure: Boolean = !isSuccess
+        def isFailure: Boolean =
+            self.isInstanceOf[Failure[?]]
 
-        inline def fold[B](inline ifFailure: Throwable => B)(inline ifSuccess: A => B): B =
+        def isPanic: Boolean =
+            self.isInstanceOf[Panic]
+
+        inline def fold[U](inline ifFailure: E => U)(inline ifPanic: Throwable => U)(inline ifSuccess: T => U): U =
             (self: @unchecked) match
-                case self: Failure[A] =>
-                    ifFailure(self.exception)
+                case self: Failure[E] =>
+                    ifFailure(self.error)
+                case self: Panic =>
+                    ifPanic(self.exception)
                 case _ =>
                     try
                         self match
-                            case self: SuccessFailure[A] @unchecked =>
-                                ifSuccess(self.unnest.asInstanceOf[A])
+                            case self: SuccessFailure[T] @unchecked =>
+                                ifSuccess(self.unnest.asInstanceOf[T])
                             case self =>
-                                ifSuccess(self.asInstanceOf[A])
+                                ifSuccess(self.asInstanceOf[T])
                     catch
-                        case ex if NonFatal(ex) => ifFailure(ex)
+                        case ex if NonFatal(ex) => ifPanic(ex)
 
-        inline def get: A =
-            fold(ex => throw ex)(identity)
+        def get(using E =:= Nothing): T =
+            fold(ex => throw new NoSuchElementException(s"Cause: $ex"))(throw _)(identity)
 
-        inline def getOrElse[B >: A](inline default: => B): B =
-            fold(_ => default)(identity)
+        inline def getOrElse[U >: T](inline default: => U): U =
+            fold(_ => default)(_ => default)(identity)
 
-        def orElse[B >: A](alternative: => Result[B]): Result[B] =
-            fold(_ => alternative)(v => Result.success(v))
+        def orElse[E2, U >: T](alternative: => Result[E2, U]): Result[E | E2, U] =
+            fold(_ => alternative)(_ => alternative)(v => Result.success(v))
```

Comment:
minor: I prefer function syntax where possible - hopefully it would desugar the same anyways:
`fold(_ => alternative)(_ => alternative)(Result.success)`